app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClient.java

```diff
@@ -0,0 +1,459 @@
+                     * the countdown.
+                     */
+                    TIMER.cancel();
+                    TIMER = new Timer();
+                    TIMER.schedule(new Task(), WAIT_MS);
+                }
+            }
+        }
+
+        /**
+         * If sending is enabled, attempt to send the provided events.
+         *
+         * Since different streams may have different configured destination URLs, events are first
+         * grouped by stream name, then sent in group-by-group batches.
+         *
+         * @param events enqueued events
+         */
+        static void send(List<Event> events) {
+            if (!ENABLED) {
+                return;
+            }
+            Map<String, ArrayList<Event>> eventsByStream = new HashMap<>();
+            for (Event event : events) {
+                String stream = event.getStream();
+                if (!eventsByStream.containsKey(stream) || eventsByStream.get(stream) == null) {
+                    eventsByStream.put(stream, new ArrayList<>());
+                }
+                eventsByStream.get(stream).add(event);
+            }
+            for (String stream : eventsByStream.keySet()) {
+                postEvents(getStreamConfig(stream), eventsByStream.get(stream));
+            }
+        }
+
+    }
+
+    /* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+    /**
+     * AssociationController: provides associative identifiers and manage their
+     * persistence
+     *
+     * Identifiers correspond to various scopes e.g. 'pageview', 'session', and 'device'.
+     *
+     * TODO: Possibly get rid of the pageview type?  Does it make sense on apps?  It is not in the iOS library currently.
+     * On apps, a "session" starts when the app is loaded, and ends when completely closed, or after 15 minutes of inactivity
+     * Save a ts when going into bg, then when returning to foreground, & if it's been more than 15 mins, start a new session, else continue session from before
+     * Possible to query/track time since last interaction? (For future)
+     */
+    static class AssociationController {
+        private static String PAGEVIEW_ID = null;
+        private static String SESSION_ID = null;
+
+        /**
+         * Generate a pageview identifier.
+         *
+         * @return pageview ID
+         *
+         * The identifier is a string of 20 zero-padded hexadecimal digits
+         * representing a uniformly random 80-bit integer.
+         */
+        static String getPageViewId() {
+            if (PAGEVIEW_ID == null) {
+                PAGEVIEW_ID = generateRandomId();
+            }
+            return PAGEVIEW_ID;
+        }
+
+        /**
+         * Generate a session identifier.
+         *
+         * @return session ID
+         *
+         * The identifier is a string of 20 zero-padded hexadecimal digits
+         * representing a uniformly random 80-bit integer.
```

Comment:
Comment for discussion: why don't why use a (v4) UUID for session IDs (across all clients)?  It seems to be the convention for all other randomly generated IDs, and would probably also be easier and better ensure consistency across platforms.

Reply:
I stumbled upon a partial answer to my own question. This token format is following that produced by [mw.user.generateRandomSessionId](https://github.com/wikimedia/mediawiki/blob/master/resources/src/mediawiki.user.js#L25-L75). There was some discussion around how this ID is generated in [T240452](https://phabricator.wikimedia.org/T240452).  I still think we'd be better off joining the rest of the world and using a [UUID](https://www.npmjs.com/package/uuid) for this rather than a custom ID format. (Would changing it break anything? If I understand correctly, this is only used for event logging.)

Reply:
+1 for using something that is used in virtually all other contexts where a "unique identifier" is needed.

Reply:
+1 to this. Just want to note for later:

We may want to create a `mw.user.generateUUIDv4()` (since [`generateRandomSessionId` is in use by a bunch of instruments](https://github.com/search?q=org%3Awikimedia+mw.user.generateRandomSessionId&type=code) that then use it to determine sampling) and figure out a migration plan for instruments that use `mw.user.sessionId()` (which relies on `mw.user.generateRandomSessionId()`)

Reply:
Excellent.  Seems like we're all on the same page on this.

I imagine it would create little to no migration work for existing consumers if we just have `mw.user.generateRandomSessionId` itself call out to `mw.user.generateUUIDv4`.  If we do that, then I think the only thing we'd need to do is announce the change a little while beforehand in case anyone is doing anything weird like relying on the token format.

Reply:
Phabbed: https://phabricator.wikimedia.org/T266813